### PR TITLE
Move default admin

### PR DESCRIPTION
**Pull request by Firesphere**
Required for https://github.com/silverstripe/silverstripe-framework/issues/6387

### CODE REVIEW

File: src/Security/MemberAuthenticator/MemberAuthenticator.php [hunk 1]
```diff
@@ -53,68 +57,99 @@ public function authenticate($data, &$result = null)
...
             $member = Member::get()
-                ->filter([Member::config()->get('unique_identifier_field') => $email])
+                ->filter([$identifierField => $email])
                 ->first();
         }
 
         // Validate against member if possible
         if ($member && !$asDefaultAdmin) {
-            $result = $member->checkPassword($data['Password']);
+            $member->checkPassword($data['Password'], $result);
```
Comment (Firesphere) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: $this->checkPassword()

File: src/Security/Member.php [hunk 2]
```diff
@@ -350,8 +308,17 @@ public function checkPassword($password)
      */
     public function isDefaultAdmin()
     {
-        return Security::has_default_admin()
-            && $this->Email === Security::default_admin_username();
+        return DefaultAdminService::isDefaultAdmin($this->Email);
```
Comment (Firesphere) [inline src/Security/Member.php]: Should have a deprecation notice

File: src/Security/MemberAuthenticator/MemberAuthenticator.php [hunk 3]
```diff
@@ -53,68 +57,108 @@ public function authenticate($data, &$result = null)
...
+     *
+     * Note, we don't return early, to prevent differences in timings to give away if a member
+     * password is invalid.
+     *
+     * @param Member $member
+     * @param string $password
+     * @param ValidationResult $result
+     * @return ValidationResult
+     */
+    public function checkPassword(Member $member, $password, ValidationResult $result = null)
```
Comment (Firesphere) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: $result is immediately overwritten

**Comment by Firesphere** [comment]
Aside from a little change I requested on Damian's update, this looks good for me to go and would wrap up everything on possible Authentication improvements.

File: src/Security/Member.php (hunk 2 above)
Comment (tractorcow) [inline src/Security/Member.php]: It's probably ok.

File: src/Security/MemberAuthenticator/MemberAuthenticator.php (hunk 3 above)
Comment (tractorcow) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: It's also passed into validateCanLogin() before being overwritten, so it's safe.

File: src/Security/MemberAuthenticator/MemberAuthenticator.php [hunk 4]
```diff
@@ -62,26 +62,24 @@ public function authenticate($data, &$result = null)
...
-            //protect against failed login
-            if ($validAdmin->isValid() && $result->isValid()) {
-                return $member;
-            } else {
-                $result->addError(_t(
-                    'SilverStripe\\Security\\Member.ERRORWRONGCRED',
-                    "The provided details don't seem to be correct. Please try again."
-                ));
+            $member = DefaultAdminService::singleton()->findOrCreateDefaultAdmin();
+            $member->validateCanLogin($result);
```
Comment (flamerohr) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: `$result = $member->validateCanLogin($result);`?
I'm not sure if this call does anything otherwise

Comment (Firesphere) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: I believe it should indeed be that.

**Comment by flamerohr** [comment]
After Damian makes the signature update, I'm happy to merge (or he can if it's green before I get home)

**Comment by tractorcow** [comment]
Ok, pushed up fix with your feedback @flamerohr

File: src/Security/MemberAuthenticator/MemberAuthenticator.php (hunk 4 above)
Comment (flamerohr) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: Damian has refactored `validateCanLogin(&$result)` which was pretty much what I had asked :D a consistent method signature

Comment (Firesphere) [inline src/Security/MemberAuthenticator/MemberAuthenticator.php]: Ah, right. Agreed, I'm good :)

File: sh.exe.stackdump [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Stack trace:
```
Comment (dhensby) [inline sh.exe.stackdump]: Was this file really needed ;)

Comment (flamerohr) [inline sh.exe.stackdump]: hey...! did not notice it :(

Comment (dhensby) [inline sh.exe.stackdump]: Hmm - it looks like it's been removed at some point from the main repo - so all goo

Comment (tractorcow) [inline sh.exe.stackdump]: I removed it manually, sorry for the mistake!

Comment (Firesphere) [inline sh.exe.stackdump]: That was my mistake, swiftly pushing because Damian wanted to take over. Sorry!

Comment (dhensby) [inline sh.exe.stackdump]: no problem - it's gone now so allllll good. :)
